CarpetX/src/io_tsv.cxx

```diff
@@ -209,7 +244,27 @@ void WriteTSVArrays(const cGH *restrict cctkGH, const std::string &filename,
       file << sep << (dir == out_dir ? i : 0);
     const int tl = 0;
     for (int vi = 0; vi < arraygroupdata.numvars; ++vi)
-      file << sep << arraygroupdata.data.at(tl).at(np * vi + DI[out_dir] * i);
+      switch (cgroup.vartype) {
+      case CCTK_VARIABLE_REAL:
+        file << sep
+             << *(CCTK_REAL *)arraygroupdata.data.at(tl).data_at(
+                    np * vi + DI[out_dir] * i);
+        break;
+      case CCTK_VARIABLE_INT:
+        file << sep
+             << *(CCTK_INT *)arraygroupdata.data.at(tl).data_at(
```

Comment:
The expression `arraygroupdata.data.at(tl).data_at(np * vi + DI[out_dir] * i)` could be calculated before the `switch` statement to reduce code duplication.

Reply:
It could, something like this:  
\`\`\`  
    for (int vi = 0; vi \< arraygroupdata.numvars; ++vi)  
     file \<\< sep;  
     const void \*ptr = arraygroupdata.data.at(tl).data\_at(np \* vi + DI\[out\_dir\] \* i);  
     switch (cgroup.vartype) {  
     case CCTK\_VARIABLE\_REAL:  
       file \<\< \*(const CCTK\_REAL \*)ptr;  
       break;  
     case CCTK\_VARIABLE\_INT:  
       file \<\< \*(const CCTK\_INT \*)ptr;  
       break;  
     case CCTK\_VARIABLE\_COMPLEX: {  
       CCTK\_COMPLEX value = \*(const CCTK\_COMPLEX \*)ptr;  
       file \<\< value.real() \<\< sep \<\< value.imag();  
     } break;  
     default:  
       assert(0 && "Unexpected variable type");  
       break;  
     }  
   file \<\< "\\n";  
 }  
\`\`\`  
though it comes at the price of information about what is output being less local and requiring backtracking in the code when reading it.

Reply:
In addition to `AnyVector`, you could introduce an `AnyScalar`, and add a virtual function that outputs to a stream. (Not sure whether that makes the code simpler or more complex in the end.)

Reply:
Hmm, maybe. Not having all those `switch` scattered around would be nice.   
  
The tricky one is the `CCTK_COMPLEX` output since it currently uses `sep` which would not be known to `AnyScalar::operator<<` .  
  
Still an `AnyScalar` and a `AnyScalar AnyVector::operator[](size_t i) const` might be useful. Would be a return by value (at least unless I want to really complicate my life and have the AnyScalar carry a back-reference to its parent `AnyVector` so that assignment to it would change the `AnyVector` element) for now.

Reply:
Something like 98ab37c ? So far only used for output. Could possibly also be used for setting poison values (though there performance issues may show up).